src/theory/quantifiers/sygus/sygus_utils.cpp

```diff
@@ -0,0 +1,182 @@
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief generic sygus utilities
+ **/
+
+#include "theory/quantifiers/sygus/sygus_utils.h"
+
+#include "expr/node_algorithm.h"
+#include "theory/quantifiers/quantifiers_attributes.h"
+#include "theory/quantifiers/sygus/sygus_grammar_cons.h"
+
+using namespace CVC4::kind;
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+/**
+ * Attribute for specifying a solution for a function-to-synthesize. This is
+ * used for marking certain functions that we have solved for, e.g. during
+ * preprocessing.
+ */
+struct SygusSolutionAttributeId
+{
+};
+typedef expr::Attribute<SygusSolutionAttributeId, Node> SygusSolutionAttribute;
+
+Node SygusUtils::mkSygusConjecture(const std::vector<Node>& fs,
+                                   Node conj,
+                                   const std::vector<Node>& iattrs)
+{
+  Assert(!fs.empty());
+  NodeManager* nm = NodeManager::currentNM();
+  std::vector<Node> ipls;
+  SygusAttribute ca;
+  Node sygusVar = nm->mkSkolem("sygus", nm->booleanType());
+  sygusVar.setAttribute(ca, true);
+  Node instAttr = nm->mkNode(INST_ATTRIBUTE, sygusVar);
+  ipls.push_back(instAttr);
+  // insert the remaining instantiation attributes
+  ipls.insert(ipls.end(), iattrs.begin(), iattrs.end());
+  Node ipl = nm->mkNode(INST_PATTERN_LIST, ipls);
+  Node bvl = nm->mkNode(BOUND_VAR_LIST, fs);
+  return nm->mkNode(FORALL, bvl, conj, ipl);
+}
+
+Node SygusUtils::mkSygusConjecture(const std::vector<Node>& fs, Node conj)
+{
+  std::vector<Node> iattrs;
+  return mkSygusConjecture(fs, conj, iattrs);
+}
+
+Node SygusUtils::mkSygusConjecture(const std::vector<Node>& fs,
+                                   Node conj,
+                                   const Subs& solvedf)
+{
+  Assert(!fs.empty());
+  NodeManager* nm = NodeManager::currentNM();
+  std::vector<Node> iattrs;
+  // take existing properties, without the previous solves
+  SygusSolutionAttribute ssa;
+  // add the current solves, which should be a superset of the previous ones
+  for (size_t i = 0, nsolved = solvedf.size(); i < nsolved; i++)
+  {
+    Node eq = solvedf.getEquality(i);
+    Node var = nm->mkSkolem("solved", nm->booleanType());
+    var.setAttribute(ssa, eq);
+    Node ipv = nm->mkNode(INST_ATTRIBUTE, var);
+    iattrs.push_back(ipv);
+  }
+  return mkSygusConjecture(fs, conj, iattrs);
+}
+
+void SygusUtils::decomposeSygusConjecture(Node q,
+                                          std::vector<Node>& fs,
+                                          std::vector<Node>& unsf,
+                                          Subs& solvedf)
+{
+  Assert(q.getKind() == FORALL);
+  Assert(q.getNumChildren() == 3);
+  Node ipl = q[2];
+  Assert(ipl.getKind() == INST_PATTERN_LIST);
```

Comment:
Maybe we wanna rename this at some point? The attribute is used for more than that.

Reply:
Yes, this kind should be renamed (not this PR of course).